keps/sig-storage/3476-volume-group/README.md

```diff
@@ -0,0 +1,1680 @@
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: pvc1
+spec:
+  accessModes:
+  - ReadWriteOnce
+  dataSource: null
+  resources:
+	requests:
+  	storage: 1Gi
+  storageClassName: storageClass1
+  volumeMode: Filesystem
+  volumeGroupNames: [volumeGroup1]
+```
+
+#### Create VolumeGroupSnapshot
+
+Create a VolumeGroupSnapshotClass:
+```
+apiVersion: volumegroup.storage.k8s.io/v1alpha1
+kind: VolumeGroupSnapshotClass
+metadata:
+  name: volumeGroupSnapshotClass1
+spec:
+  parameters:
+     …...
+```
+
+A VolumeGroupSnapshot taken from the VolumeGroup dynamically:
+```
+apiVersion: volumegroup.storage.k8s.io/v1alpha1
+kind: VolumeGroupSnapshot
+metadata:
+  name: my-group-snapshot
+spec:
+  source:
+      volumeGroupName: volumeGroup1
+  volumeGroupSnapshotClassName: volumeGroupSnapshotClass1
+```
+
+A new external VolumeGroup controller will handle VolumeGroupClass, VolumeGroup, and VolumeGroupContent resources. We may need to split this into two controllers, one common controller that handles common functions such as binding, and one sidecar controller that calls the CSI driver.
+
+External provisioner will be modified to read information from volume groups (through volumeGroupNames) and pass them down to the CSI driver.
+
+A new external VolumeGroupSnapshot controller will handle VolumeGroupSnapshotClass, VolumeGroupSnapshot, and VolumeGroupSnapshotContent resources. We may need to split this into two controllers, one common controller that handles common functions such as binding, and one sidecar controller that calls the CSI driver.
+
+Snapshot controller will be modified to update VolumeSnapshot status. External snapshotter sidecar will be modified to update VolumeSnapshotContent status.
+
+### CSI Changes
+
+#### CSI Capabilities
+
+New controller capabilities CREATE_DELETE_VOLUME_GROUP, VOLUME_GROUP_ADD_REMOVE_EXISTING_VOLUME, CREATE_DELETE_GROUP_SNAPSHOT, INDIVIDUAL_SNAPSHOT_RESTORE, GET_VOLUME_GROUP, GET_VOLUME_GROUP_SNAPSHOT, LIST_VOLUME_GROUPS, LIST_VOLUME_GROUP_SNAPSHOTS will be added.
+
+* CREATE_DELETE_VOLUME_GROUP:
+  Indicates that the controller plugin supports creating and deleting a volume group.
+
+* VOLUME_GROUP_ADD_REMOVE_EXISTING_VOLUME:
+  Indicates that the controller plugin supports adding an existing volume to a volume
+  group and removing a volume from a volume group without deleting it.
+
+* CREATE_DELETE_GROUP_SNAPSHOT:
+  Indicates that the controller plugin supports creating a snapshot of all volumes
+  in a volume group.
+
+* INDIVIDUAL_SNAPSHOT_RESTORE:
+  Indicates whether the controller plugin supports creating a volume from an
+  individual volume snapshot if the volume snapshot is part of a
+  VolumeGroupSnapshot. Use cases: selective restore, advanced recovery, etc.
+  Note: In Phase 1, this is the only way to restore after taking a group snapshot.
+  User can create a volume from a volume snapshot for all the individual snapshots
+  created along the group snapshot.
+
+* GET_VOLUME_GROUP:
```

Comment:
Is this required for some storage systems? Generally get/list is a much cheaper call than create and will have higher api quota.

Also using create as a "get" messes up our csi metrics, because it shows "create" getting called very frequently with low latency: https://github.com/kubernetes-csi/external-snapshotter/pull/165#discussion_r318373725

Reply:
I agree that Get is a cheaper call than Create. We'll use Get to check status if it is supported by the driver. Are you suggesting that we make "Get" a required call?  We probably can't do that.